terror/terror.go

```diff
@@ -0,0 +1,291 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package terror
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"go.uber.org/zap"
+)
+
+// ErrCode represents a specific error type in a error class.
+// Same error code can be used in different error classes.
+type ErrCode int
+
+// ErrCodeText is a textual error code that represents a specific error type in a error class.
+// This string should include component name, see
+// https://github.com/pingcap/tidb/blob/master/docs/design/2020-05-08-standardize-error-codes-and-messages.md#the-error-code-range
+type ErrCodeText string
+
+// ErrClass represents a class of errors.
+type ErrClass int
+
+var errClass2Desc = make(map[ErrClass]string)
+
+// RegisterErrorClass registers new error class for terror.
+func RegisterErrorClass(classCode int, desc string) ErrClass {
+	errClass := ErrClass(classCode)
+	if _, exists := errClass2Desc[errClass]; exists {
+		panic(fmt.Sprintf("duplicate register ClassCode %d - %s", classCode, desc))
+	}
+	errClass2Desc[errClass] = desc
+	return errClass
+}
+
+// String implements fmt.Stringer interface.
+func (ec ErrClass) String() string {
+	if s, exists := errClass2Desc[ec]; exists {
+		return s
+	}
+	return strconv.Itoa(int(ec))
+}
+
+// EqualClass returns true if err is *Error with the same class.
+func (ec ErrClass) EqualClass(err error) bool {
+	e := errors.Cause(err)
+	if e == nil {
+		return false
+	}
+	if te, ok := e.(*Error); ok {
+		return te.class == ec
+	}
+	return false
+}
+
+// NotEqualClass returns true if err is not *Error with the same class.
+func (ec ErrClass) NotEqualClass(err error) bool {
+	return !ec.EqualClass(err)
+}
+
+// New defines an *Error with an error code and an error message.
+// Usually used to create base *Error.
+// Currently, this method is same as Synthesize after extracted from parser.
+// deprecated: textual error codes is more readable than numeric error codes, use NewError instead.
+func (ec ErrClass) New(code ErrCode, message string) *Error {
+	return &Error{
+		class:   ec,
+		code:    code,
+		message: message,
+	}
+}
+
+// NewError defines an *Error with an error code, its code message and an error message.
+//
+func (ec ErrClass) NewError(code ErrCode, text ErrCodeText, message string) *Error {
+	return &Error{
+		class:    ec,
+		code:     code,
+		codeText: text,
+		message:  message,
+	}
+}
+
+// Synthesize synthesizes an *Error in the air
+// it didn't register error into ErrClassToMySQLCodes
+// so it's goroutine-safe
+// and often be used to create Error came from other systems like TiKV.
+func (ec ErrClass) Synthesize(code ErrCode, message string) *Error {
+	return &Error{
+		class:   ec,
+		code:    code,
+		message: message,
+	}
+}
+
+// Error implements error interface and adds integer Class and Code, so
+// errors with different message can be compared.
+type Error struct {
+	class ErrClass
```

Comment:
How does this field map to RFC https://github.com/pingcap/tidb/blob/master/docs/design/2020-05-08-standardize-error-codes-and-messages.md?